kernel/libc/koslib/gmon.c

```diff
@@ -0,0 +1,386 @@
+#define SAMPLE_FREQ            100 /* 100 Hz (10 ms intervals) */
+
+/* The type that holds the count for the histogram every 10 ms */
+#define	HIST_COUNTER_TYPE      uint16_t
+
+/* Each histogram counter represents 16 bytes (4 instructions) */
+#define HISTFRACTION  4
+
+/* Practical value for all archs */
+#define	HASHFRACTION  2
+
+/* Percentage of the text size we want to allocate for our BST */
+#define ARCDENSITY	  2
+
+/* GMON file header */
+typedef struct gmon_hdr {
+    uintptr_t lowpc;
+    uintptr_t highpc;
+    size_t ncnt;       /* size of histogram buffer (plus this header) */
+    uint32_t version;  /* version number; 0x00051879 */
+    uint32_t profrate; /* profiling clock rate */
+    uint32_t spare[3]; /* reserved */
+} gmon_hdr_t;
+
+typedef struct gmon_node {
+    uintptr_t selfpc; /* callee address/program counter. The caller address is in froms[] array which points to tos[] array */
+    uint32_t count; /* how many times it has been called */
+    uint16_t left; /* link to next entry in hash table. For tos[0] this points to the last used entry */
+    uint16_t right; /* additional padding bytes, to have entries 4byte aligned */
+} gmon_node_t;
+
+/* GMON arc */
+typedef struct gmon_arc {
+    uintptr_t frompc;
+    uintptr_t selfpc;
+    size_t count;
+} gmon_arc_t;
+
+/* GMON profiling states */
+typedef enum gmon_state {
+    GMON_PROF_ON,
+    GMON_PROF_BUSY,
+    GMON_PROF_ERROR,
+    GMON_PROF_OFF
+} gmon_state_t;
+
+/* GMON context */
+typedef struct gmon_context {
+    gmon_state_t state;
+    uintptr_t lowpc;
+    uintptr_t highpc;
+    uintptr_t pc;
+    size_t textsize;
+
+    size_t nfroms;
+    uint16_t *froms;
+
+    size_t nnodes;
+    gmon_node_t *nodes;
+    size_t node_count;
+
+    size_t ncounters;
+    HIST_COUNTER_TYPE *histogram;
+
+    kthread_t *histogram_thread;
+
+    volatile bool running_thread;
+    volatile bool initialized;
+} gmon_context_t;
+
+static gmon_context_t g_context = {
+    .state = GMON_PROF_OFF,
+    .lowpc = 0,
+    .highpc = 0,
+    .pc = 0,
+    .textsize = 0,
+    .nfroms = 0,
+    .froms = NULL,
+    .nnodes = 0,
+    .nodes = NULL,
+    .node_count = 0,
+    .ncounters = 0,
+    .histogram = NULL,
+    .histogram_thread = NULL,
+    .running_thread = false,
+    .initialized = false
+};
+
+/* Used to write the arcs */
+static void traverse_and_write(FILE *out, gmon_context_t *cxt, uint32_t index, uintptr_t from_addr) {
+    gmon_arc_t arc;
+    gmon_node_t *node;
+
+    if(index == 0)
+        return;
+    
+    /* Grab a node */
+    node = &cxt->nodes[index];
+
+    /* Traverse the left subtree */
+    traverse_and_write(out, cxt, node->left, from_addr);
+
+    /* Write the arc */
+    arc.frompc = from_addr;
+    arc.selfpc = node->selfpc;
+    arc.count = node->count;
+    fwrite(&arc, sizeof(gmon_arc_t), 1, out);
+
+    /* Traverse the right subtree */
+    traverse_and_write(out, cxt, node->right, from_addr);
+}
+
+static void *histogram_callback(void *arg) {
```

Comment:
Mind naming this something like `histogram_thread(..)` or `histogram_exec(..)` or something to distinguish between just a one-off callback and something that is a persistent thread callback which isn't just immediately returning?

Reply:
Didn't think of it that way. Makes sense. histaogram_thread it is.